R/workflow.R

```diff
@@ -0,0 +1,76 @@
+##' Run a workflow.
+##'
+##' WIP - skeleton implementation
+##'
+##' @param name Name of workflow to run
+##'
+##' @inheritParams orderly_list
+##'
+##' @return ID of workflow run
+##' @export
+##'
+##' @examples
+##' path <- orderly::orderly_example("demo")
+##'
+##' # To run most reports, provide the report name (and the path if
+##' # not running in the working directory, as is the case here):
+##' id <- orderly::orderly_workflow("my_workflow", root = path)
+orderly_workflow <- function(name, root = NULL, locate = TRUE) {
+  ## Locate file
+  config <- orderly_config_get(root, locate)
+  workflow <- orderly_workflow_get(name, config)
+
+  ids <- workflow$run()
+
+  ## TODO: Write to DB after run
+
+  ids
+}
+
+orderly_workflow_get <- function(name, config) {
+  filename <- path_orderly_workflow_dir(config$root, name)
+  assert_file_exists(basename(filename), workdir = dirname(filename),
+                     name = "workflow configuration")
+  workflow$new(name, filename, config)
+}
+
+workflow <- R6::R6Class(
+  "workflow",
+
+  public = list(
+    workflow_name = NULL,
+    steps = NULL,
+    workflow_id = NULL,
+    config = NULL,
+
+    initialize = function(workflow_name, workflow_path, config) {
+      self$workflow_name <- workflow_name
+      self$workflow_id <- ids::random_id()
+      self$config <- config
+      raw <- yaml_read(workflow_path)
+      self$steps <- parse_steps(raw)
+      validate_steps(self$steps, self$config, workflow_name)
+    },
+
+    run = function() {
+      ## TODO: Implement run
+      orderly_log("workflow", paste0("Running workflow ", self$workflow_id))
+      ids::random_id()
+    }
+  )
+)
+
+parse_steps <- function(yml) {
+  unname(unlist(yml$steps))
+}
+
+validate_steps <- function(steps, config, workflow_name) {
+  reports <- basename(list_dirs(path_src(config$root)))
```

Comment:
This is the sample impl as `orderly_list` but using the config object. Looks like the impl of the `orderly_list` stuff if half way through a move to a `list2` function so wanted to mention this. Don't know whether it is worth adding an `orderly_list(config)` function which is called both here and in `orderly_list(root, locate)`? Or some other way of tidying up the code re-use

Reply:
yeah, there's a bunch of stuff like this that needs tidying up. I'm happy for you to tidy up here or in another ticket. It's internal stuff only so we can do it however we want

Reply:
Or you can just use `orderly_list(config)` which works fine as `orderly_config_get(config)` just returns `config`

Reply:
Oh nice - I missed that. `orderly_list(config)` is perfect.